### ISSUE TITLE
Need first-class release documentation generation code

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2925]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#2925|
|Type|enhancement|
|Created|2007-11-24 03:55:27Z|
|Branch|https://github.com/twisted/twisted/tree/generate-release-docs-2925|

admin/process-docs and the lore-calling code in admin/release-twisted are both nightmares. Let's do something TDD which generates our documentation for release in a non-sucky way.

<details><summary>Searchable metadata</summary>

```
trac-id__2925 2925
type__enhancement enhancement
reporter__radix radix
priority__normal normal
milestone__twisted_8_0 twisted-8.0
branch__branches_generate_release_docs_2925 branches/generate-release-docs-2925
branch_author__ 
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1195876527000000 1195876527000000
changetime__1198861768000000 1198861768000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix commented|
|-|-|

(In [21957]) Branching to 'generate-release-docs-2925'

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_2"></a>|@radix commented|
|-|-|

Status update: It's got pretty much the basic "lore" functionality that release-twisted has, but it still needs to do the rest of the stuff that admin/process-docs does, like generating man page HTML and generating the book.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix removed owner|
|-|-|

Ok, I guess this can be reviewed, maybe.

It doesn't generate book stuff; it only does what's necessary to facilitate the tarball building so far. I've created other tickets ([#2940](https://github.com/twisted/twisted/issues/2940), [#2941](https://github.com/twisted/twisted/issues/2941)) for that stuff.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @radix<br>@exarkun set status to `new`|
|-|-|


  * twisted/python/test/test_release.py
    * import section uses backslash line continuations against the coding standard
    * ```@ivar```s in ```DocBuilderTestCase.setUp``` are weird - what does pydoctor do with these?
    * ```lore_docs``` doesn't follow naming convention
    * The two-pass interpolation in ```getArbitraryLoreInput``` is a bit more confusing than the equivalent single pass implementation would be.  Any particular reason you did it this way?
    * These byte-exact xhtml tests are going to hurt for future lore maintenance.  I don't have a suggestion for another way to do this, though.  Maybe a comment or something pointing out that the exact bytes aren't really important, just the fact that lore ends up doing the right kind of processing.
  * twisted/python/_release.py
    * ```DocBuilder.build``` has a local ```input_files``` which doesn't follow the naming convention
    * Why nest the lore input inside ```DocBuilder.build```?
    * link generation (ie, the 4 lines after the --linkrel comment) could be a separate method and have unit tests.  This might not remove the need for any of the byte-exact xhtml assertions in the test module, but it would at least allow lore-independent unit tests for that logic.
    * ```filePathDelta```'s arguments would seem to be better named ```origin``` and ```destination```.
    * ```list.extend``` is a nice method.  So is ```list.__add__```, even.  Can you use one of those instead of ```list.__iadd__``` at the end of ```filePathDelta```?
  * twisted/python/dist.py
    * FilePath imported but unused
  * twisted/python/test/test_dist.py
    * three new blank lines at the end of the file, not sure if that was intentional or not

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix set owner to @exarkun|
|-|-|

Ok, I've dealt with everything. Your questions are now answered:

 * import of lore moved to top of module, comment added
 * As we discussed on IRC, @ivar in a method documents the instance variables in that method.
 * the two-pass interpolation was for a use case that is no longer. I've got rid of the 'body' parameter entirely.
 * I've added a maintainer's note to the class docstring for the tests, explaining the byte-equality tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @radix|
|-|-|

Attribute the lore import comment in _release.py so people can resolve "me" to someone who isn't them.  Everything else seems great, no need for re-review.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_9"></a>|@radix set status to `closed`|
|-|-|

(In [22169]) Merge generate-release-docs-2925

Author: therve,radix
Reviewer: exarkun
Fixes [#2925](https://github.com/twisted/twisted/issues/2925)

A DocBuilder object is added to twisted.python._release which offers
convenience for building directories full of documentation with lore.

The functionality included is equivalent with the lore-building code
in admin/release-twisted.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
